File: StriverA2Z/C++/13_BinaryTree/top_view.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

vector<int> topView(Node *root) {
    vector<int> ans;
    if (!root) return ans;
    queue<pair<Node *, int>> q;
    map<int, int> mpp;
    q.push({root, 0});
    while (!q.empty()) {
        Node *node = q.front().first;
        int level = q.front().second;
        q.pop();
        if (mpp.find(level) == mpp.end()) mpp[level] = node->data;
        if (node->left) {
            q.push({node->left, level - 1});
        }
        if (node->right) {
            q.push({node->right, level + 1});
        }
    }
    for (auto it: mpp) {
        ans.push_back(it.second);
    }
    return ans;
}

int main() {
    Node *root = arrayToBTLevelOrder({1, 2, 3, 4, 5, 6, 7});
    vector<int> ans = topView(root);
    for (int i: ans) {
        cout << i << " ";
    }
    return 0;
}

File: StriverA2Z/Java/BinaryTrees/TopView.java
import java.util.*;

public class TopView {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6, 7});
        List<Integer> ans = topView(root);
        for (Integer i : ans) {
            System.out.print(i + " ");
        }
    }
    public static List<Integer> topView(Node root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        Queue<Pair> q = new LinkedList<>();
        TreeMap<Integer, Integer> mpp = new TreeMap<>();
        q.offer(new Pair(root, 0));
        while (!q.isEmpty()) {
            Pair p = q.poll();
            Node node = p.node;
            int level = p.num;
            if (!mpp.containsKey(level)) mpp.put(level, node.data);
            if (node.left != null) q.offer(new Pair(node.left, level -1));
            if (node.right != null) q.offer(new Pair(node.right, level + 1));
        }
        for (Map.Entry<Integer, Integer> i : mpp.entrySet()) {
            ans.add(i.getValue());
        }
        return ans;
    }
}
